packages/serverpod/lib/src/util/database_util.dart

```diff
@@ -0,0 +1,81 @@
+import 'package:serverpod/serverpod.dart';
+
+/// Provides utility functions for working with the [Database].
+abstract class DatabaseUtil {
+  /// Runs the closure [f] in isolation and undoes any changes to the database in case [f] fails.
+  ///
+  /// Depending on whether the [transaction] parameter refers to an active transaction (non-`null`)
+  /// or not, the method has 2 slightly distinct behaviors:
+  ///
+  /// - In case no [transaction] was given (`null`), the method [f] is run in a new transaction.
+  ///
+  ///   If [f] should error, the internally created transaction is cancelled.
+  ///
+  ///   Code outside of [f] accessing the database, which does not have access to the internal
+  ///   transaction, can not see any changes until this method finishes.
+  ///
+  ///   Once this method is finished, all changes are persisted to the database.
+  ///
+  /// - In case an active [transaction] was given, this method creates a [savepoint](https://www.postgresql.org/docs/current/sql-savepoint.html)
```

Comment:
This is quite a mouthful.

For our auth cases this is probably fine.

For external users I wonder whether we should ensure (on the lower-level methods already?) that there is not parallel savepoint ongoing, so that there is only 1 thing to be undone (unless the user would explicitly opt into "unsafe" nested / parallel savepoints).

Reply:
To be pondered after more experience using the helper.